### PR TITLE
Make fork enableable from config.json

### PR DESCRIPTION
**Pull request by paulcallen**
Signed-off-by: Paul Allen <paul.c.allen@microsoft.com>

### CODE REVIEW

File: doc/design/fork.md
Comment (mikbras) [inline doc/design/fork.md:21]: Typo: should be "parent's"

File: doc/design/fork.md
Comment (jxyang) [inline doc/design/fork.md:29]: It's not clear to me for the first 2 options, whether the parent will continue when the child is running. And I am not sure they will be useful.

File: doc/design/fork.md
Comment (mikbras) [inline doc/design/fork.md:23]: Typo: /teh/the/

File: include/myst/kernel.h
Comment (mikbras) [inline include/myst/kernel.h:19]: Will you need ``myst_fork_wait_for_exec`` (where the parent waits until the ``exec()`` has started?

File: doc/design/fork.md
Comment (mikbras) [inline doc/design/fork.md:29]: Is there a corresponding ``--fork-mode`` option for ``myst exec``?

File: kernel/syscall.c
Comment (mikbras) [inline kernel/syscall.c:3945]: Seeing this makes me think the ``struct myst_fork_info`` struct as a single parameter might be cleaner.

File: kernel/syscall.c
Comment (mikbras) [inline kernel/syscall.c:3936]: Could you break this out into a function? It is getting kind of big.

File: crt/exit.c
Comment (mikbras) [inline crt/exit.c:32]: This should be ``myst_fork_mode_t`` pointer because ``sizeof(myst_fork_mode_t) != sizeof(long)``

Comment (mikbras) [reply crt/exit.c:32]: This is a pointer error.

File: tests/fork/config-none.json
Comment (mikbras) [inline tests/fork/config-none.json:16]: Is this the default if the option is omitted?

File: tests/ltp/Makefile
Comment (jxyang) [inline tests/ltp/Makefile:47]: Without `MYST_PSEUDO_FORK`, applications can still take a ForkMode option but just ignore it.

File: tests/ltp/Makefile
Comment (jxyang) [inline tests/ltp/Makefile:39]: Do we still need Mike's ltp repo? If not delete these lines.

File: tests/fork/config-wait-exec.json
Comment (mikbras) [inline tests/fork/config-wait-exec.json:16]: The strings here don't match what is in ``myst_fork_mode_t`` (I saw ``myst_fork_pseudo_wait_for_exit_exec`` but not ``myst_fork_pseudo_wait_for_exec``).

File: tools/myst/config.c
Comment (mikbras) [inline tools/myst/config.c:245]: These labels do not match: ``myst_fork_pseudo_wait_for_exit_exec`` and ``pseudo_wait_for_exec``

**Review by mikbras — CHANGES_REQUESTED** [review]
Overall looks great. I found a couple of real problems and then some nits.

File: crt/Makefile
Comment (jxyang) [inline crt/Makefile:27]: We should remove `MYST_PSEUDO_FORK` altogether. Instead, default ForkMode to `none` if not specified.

Comment (paulcallen) [reply crt/Makefile:27]: kind of, but for a pretty weak reason that probably wont hold up in a discussion.
I guess once this PR goes in we should merge the LTP fork from my repo to yours, or get rid of yours.

File: include/myst/kernel.h
Comment (paulcallen) [reply include/myst/kernel.h:19]: that is the next PR

File: doc/design/fork.md
Comment (paulcallen) [reply doc/design/fork.md:29]: no because we can always use `--app-config-path config.json`

File: tests/fork/config-none.json
Comment (paulcallen) [reply tests/fork/config-none.json:16]: "ForkMode":"none"  = myst_fork_none  = 0
The config structure is either calloc-ed or malloc-ed/memset-ed

File: doc/design/fork.md
Comment (mikbras) [reply doc/design/fork.md:29]: Hmm? So far we have a complete set of options to avoid use of config.json.

File: crt/fork.c
Comment (paulcallen) [reply crt/fork.c:322]: I'll rename in this PR, but will add the structure in the next PR as I need more information anyway.

File: crt/exit.c
Comment (paulcallen) [reply crt/exit.c:32]: will do that in the next PR

File: kernel/syscall.c
Comment (paulcallen) [reply kernel/syscall.c:3936]: done

File: kernel/syscall.c
Comment (paulcallen) [reply kernel/syscall.c:3945]: next PR

File: crt/exit.c
Comment (paulcallen) [reply crt/exit.c:32]: fixed

File: tools/myst/config.c
Comment (paulcallen) [reply tools/myst/config.c:245]: done

File: tests/fork/config-wait-exec.json
Comment (paulcallen) [reply tests/fork/config-wait-exec.json:16]: renamed

File: doc/design/fork.md
Comment (paulcallen) [reply doc/design/fork.md:1]: I was thinking this document would be updated at that point to describe that and the differences between the two. I have renamed for now.

File: doc/design/fork.md
Comment (paulcallen) [reply doc/design/fork.md:3]: added an extra line.

File: doc/design/fork.md
Comment (paulcallen) [reply doc/design/fork.md:8]: done

File: doc/design/fork.md
Comment (paulcallen) [reply doc/design/fork.md:15]: done

File: doc/design/fork.md
Comment (paulcallen) [reply doc/design/fork.md:29]: There are scenarios where a non-fork exec scenario works well. The LTP test framework itself uses fork to run the tests giving it an easy way to check the return code from the tests and detecting if it crashed.

File: tests/ltp/Makefile
Comment (paulcallen) [reply tests/ltp/Makefile:47]: I have removed all references to MYST_PSEUDO_FORK define as the fork code is always compiled in, but is disabled by default.

File: tests/ltp/Makefile
Comment (paulcallen) [reply tests/ltp/Makefile:39]: removed

File: crt/Makefile
Comment (paulcallen) [reply crt/Makefile:27]: done

File: doc/design/fork.md
Comment (paulcallen) [reply doc/design/fork.md:29]: ok. adding

**Comment by mikbras** [comment]
I merged a fix for the pointer error an introduced ``myst_fork_info_t`` while I was at it.

**Review by mikbras — APPROVED** [review]
Looks good to me. Thank you for doing this Paul.

**Review by jxyang — APPROVED** [review]

**Pull request merged**
